src/main/java/net/snowflake/ingest/streaming/internal/ExternalVolumeManager.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2024 Snowflake Computing Inc. All rights reserved.
+ */
+
+package net.snowflake.ingest.streaming.internal;
+
+import static net.snowflake.ingest.connection.ServiceResponseHandler.ApiName.STREAMING_CHANNEL_CONFIGURE;
+import static net.snowflake.ingest.utils.Constants.CHANNEL_CONFIGURE_ENDPOINT;
+
+import java.io.IOException;
+import java.util.Calendar;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import net.snowflake.client.jdbc.SnowflakeSQLException;
+import net.snowflake.ingest.utils.ErrorCode;
+import net.snowflake.ingest.utils.SFException;
+
+/** Class to manage multiple external volumes */
+class ExternalVolumeManager<T> implements StorageManager<T> {
+
+  // Reference to the external volume per table
+  private final Map<String, StreamingIngestStorage> externalVolumeMap;
+  private final SnowflakeStreamingIngestClientInternal<T> owningClient;
+
+  private final boolean isTestMode;
+
+  /**
+   * Constructor for ExternalVolumeManager
+   *
+   * @param isTestMode whether the manager in test mode
+   * @param client the owning client
+   */
+  ExternalVolumeManager(boolean isTestMode, SnowflakeStreamingIngestClientInternal<T> client) {
+    this.owningClient = client;
+    this.isTestMode = isTestMode;
+    this.externalVolumeMap = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Given a blob, return the target storage by looking up the table name from the channel context
+   *
+   * @param blobData the blob to upload
+   * @return target storage
+   */
+  @Override
+  public StreamingIngestStorage getStorage(List<List<ChannelData<T>>> blobData) {
+    // Only one chunk per blob in Iceberg mode.
+    ChannelFlushContext channelContext = blobData.get(0).get(0).getChannelContext();
+    StreamingIngestStorage stage =
+        this.externalVolumeMap.get(channelContext.getFullyQualifiedTableName());
+
+    if (stage == null) {
+      throw new SFException(
+          ErrorCode.INTERNAL_ERROR,
```

Comment:
same question as another thread: does customer code get a chance to intercept this exception? Does it cause a process crash (or does it silently get ignored) ?

Reply:
This cause a process crash. I don't think this is suppose to happens given that client seems impossible to ingest to a channel without open it via `StreamingIngestClient`